### PR TITLE
Add directory-based file ownership strategy

### PR DESCRIPTION
**Pull request by pbomb**
Adds directory-based file ownership strategy. By adding a `.codeowners` file in a directory, any files in that directory and sub-directories will be owned by that team. This is a technology-agnostic approach that works with Ruby, JS, and everything else, greatly simplifying how to define ownership.

I tested this out in our company's application repository and it worked as expected.

### CODE REVIEW

**Comment by pbomb** [comment]
@shageman @alexevanczuk Can y'all take a look at this? The README has more information on this new approach.

**Comment by pbomb** [comment]
Hi @alexevanczuk.

> When would this be used? With front-end packages (I'm assuming the intent is to use this new mapping implementation in the frontend), wouldn't we want JS package ownership? Or is this for front-end code that is not a package?
I could see this being used everywhere (frontend and backend). On the frontend, it would be used for packages that have multiple owners, which is a common scenario. The Gusto backend is moving towards fewer, larger packages, so it might move to a similar situation in which related code in a package is owned by multiple, related teams. Packwerk does allow for nested packages, so that could be an approach used there, but that is not possible with NPM packages. Also, I've heard some opinions against nested packages on the backend too.

> How do you feel about the advantage of this over the owned_globs declaration in team YML files?
I vastly prefer to colocate the ownership in the filesystem rather than in some team config file somewhere else in the repository and is the primary reason why I'm much more in favor of directory-based ownership.

The other great argument for directory-based ownership (IMO), is that it is both language and package-technology agnostic. This would work in a C++, Perl, Python, or Rust codebase as well. It's similar to the package-based approach, but doesn't add our own concepts like a "metadata" object to an NPM package.json file which is not standard, for example.

File: lib/code_ownership/private/ownership_mappers/directory_ownership.rb
Comment (alexevanczuk) [inline lib/code_ownership/private/ownership_mappers/directory_ownership.rb:12]: Could we just call this `.codeowner`?

**Review by alexevanczuk — COMMENTED** [review]
Looks good. Only suggestion would be to rename the file (and reference to the file) to `.codeonwner`, since it's just one owner.

File: lib/code_ownership/private/ownership_mappers/directory_ownership.rb
Comment (pbomb) [reply lib/code_ownership/private/ownership_mappers/directory_ownership.rb:12]: Yeah, that's fine

**Comment by alexevanczuk** [comment]
Looks good, one more thing I forgot.

Could you increment the minor version? That way, once tests pass on `main`, the new gem will automatically be pushed to rubygems without any additional work.

**Comment by pbomb** [comment]
@alexevanczuk Done!

**Review by alexevanczuk — APPROVED** [review]

**Comment by alexevanczuk** [comment]
@pbomb There was an issue with CI that I just fixed on `main`. Can you rebase or merge and push and then we should be good?

**Comment by pbomb** [comment]
@alexevanczuk I merged main into this branch. I'm not sure what it's waiting on approval for?

**Comment by alexevanczuk** [comment]
The approval is since you're a first time contributor to this repo, it needs me to approve running CI (so if you were a bot you couldn't abuse CI to mine bitcoin for example). We should probably get this authz worked properly though so any Gustie has elevated permissions on `rubyatscale`.

**Comment by alexevanczuk** [comment]
@pbomb Once CI  and CD pass, you should be good to go!

**Pull request merged**
